File: algo_2021/s7.-hash-heap/129.Rehashing/129_rehashing.py
"""
Definition of ListNode
class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""
class Solution:
    """
    @param hashTable: A list of The first node of linked list
    @return: A list of The first node of linked list which have twice size
    """
    def rehashing(self, hashTable):
        # write your code here
        new_size = len(hashTable) * 2
        res = [None] * new_size
        for item in hashTable:
            while item:
                if res[item.val%new_size]:
                    self.handle_collision(item, res, new_size)
                else:
                    res[item.val%new_size] = ListNode(item.val)
                item = item.next
        return res
    
    def handle_collision(self, item, res, new_size):
        cur = res[item.val%new_size]
        while cur.next:
            cur = cur.next
        cur.next = ListNode(item.val)

File: lintcode_algo_2021/s7.-ds-design/129.Rehashing/129_rehashing.java
/**
 * Definition for ListNode
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    /**
     * @param hashTable: A list of The first node of linked list
     * @return: A list of The first node of linked list which have twice size
     */    
    public ListNode[] rehashing(ListNode[] hashTable) {
        // write your code here
        int new_size = hashTable.length * 2;
        // System.out.println(new_size);
        ListNode[] res = new ListNode[new_size];
        
        for (int i = 0; i < hashTable.length; i++) {
            while (hashTable[i] != null) {
                int index = hashTable[i].val % new_size;
                // in Java: MOD act different in negative value
                // here in hash table, -7 should be in index:3 if hash table size = 5
                // but java -7 mod 5 = -3, therefore should add table size to let it = 3
                if (index < 0) {
                    index = index + new_size;
                }
                if (res[index] != null) {
                    handleCollision(hashTable[i], res, new_size);
                } else {
                    res[index] = new ListNode(hashTable[i].val);
                }
                hashTable[i] = hashTable[i].next;
            } 
        }
        return res;
    }

    private void handleCollision(ListNode node, ListNode[] res, int new_size) {
        int index = node.val % new_size;
        if (index < 0) {
            index = index + new_size;
        }
        ListNode cur = res[index];
        while (cur.next != null) {
            cur = cur.next;
        }
        cur.next = new ListNode(node.val);
    }
};